test/performance/odp_dma_perf.c

```diff
@@ -9,266 +9,522 @@
+					 config->num_sessions, dma_capa.pool.max_pools);
+				return PRS_NOT_SUP;
+			}
+
+			if ((dma_capa.compl_mode_mask & ODP_DMA_COMPL_EVENT) == 0U) {
+				ODPH_ERR("Unsupported DMA completion mode, event\n");
+				return PRS_NOT_SUP;
+			}
+
+			if (dma_capa.queue_type_sched == 0) {
+				ODPH_ERR("Unsupported DMA queueing type, scheduled\n");
+				return PRS_NOT_SUP;
+			}
+
+			if (config->num_inflight > dma_capa.pool.max_num) {
+				ODPH_ERR("Unsupported amount of completion events: %u (max: %u)\n",
+					 config->num_inflight, dma_capa.pool.max_num);
+				return PRS_NOT_SUP;
+			}
+
+			if (odp_schedule_capability(&sched_capa) < 0) {
+				ODPH_ERR("Error querying scheduler capabilities\n");
+				return PRS_NOK;
+			}
+
+			if (config->num_sessions > sched_capa.max_groups - 3U) {
+				ODPH_ERR("Unsupported amount of scheduler groups: %u (max: %u)\n",
+					 config->num_workers, sched_capa.max_groups - 3U);
+				return PRS_NOT_SUP;
+			}
+		}
 
-	if (config->num_in_seg < 1 || config->num_in_seg > MAX_NUM_IN_SEGS) {
-		ODPH_ERR("Invalid number of input segments: %d.\n", config->num_in_seg);
-		return -1;
+		config->compl_mode_mask |= mode_map[config->compl_mode];
 	}
 
-	if (config->seg_type != TYPE_PKT && config->seg_type != TYPE_MEM) {
-		ODPH_ERR("Invalid input segment type: %d.\n", config->seg_type);
-		return -1;
+	max_trs = dma_capa.max_transfers > MAX_SEGS ? MAX_SEGS : dma_capa.max_transfers;
+
+	if (config->num_inflight == 0U)
+		config->num_inflight = max_trs;
+
+	if (config->num_inflight > max_trs) {
+		ODPH_ERR("Unsupported amount of in-flight DMA transfers: %u (max: %u)\n",
+			 config->num_inflight, max_trs);
+		return PRS_NOT_SUP;
 	}
 
-	if (check_completion_modes(config)) {
-		ODPH_ERR("Invalid completion modes.\n");
-		return -1;
+	max_in = config->num_in_segs * config->num_inflight;
+	max_out = config->num_out_segs * config->num_inflight;
+	max_segs = max_in > max_out ? max_in : max_out;
+
+	if (max_segs > MAX_SEGS) {
+		ODPH_ERR("Unsupported input/output * inflight segment combination: %u (max: %u)\n",
+			 max_segs, MAX_SEGS);
+		return PRS_NOT_SUP;
 	}
 
-	if (config->num_rounds < 1) {
-		ODPH_ERR("Invalid number of rounds: %d.\n", config->num_rounds);
-		return -1;
+	if (config->seg_type == PACKET) {
+		if (odp_pool_capability(&pool_capa) < 0) {
+			ODPH_ERR("Error querying pool capabilities\n");
+			return PRS_NOK;
+		}
+
+		if (pool_capa.pkt.max_pools < 2U) {
+			ODPH_ERR("Unsupported amount of packet pools: 2 (max: %u)\n",
+				 pool_capa.pkt.max_pools);
+			return PRS_NOT_SUP;
+		}
+
+		if (pool_capa.pkt.max_len != 0U &&
+		    config->src_seg_len > pool_capa.pkt.max_len) {
+			ODPH_ERR("Unsupported packet size: %u (max: %u)\n", config->src_seg_len,
+				 pool_capa.pkt.max_len);
+			return PRS_NOT_SUP;
+		}
+
+		if (pool_capa.pkt.max_num != 0U &&
+		    config->num_inflight * (config->num_in_segs + config->num_out_segs)
+		    > pool_capa.pkt.max_num) {
```

Comment:
Should be `MAX(num_in_segs, num_out_segs)` and also include the number of sessions?

Reply:
Yes, check seems to have missed a few rounds of updates. V4.